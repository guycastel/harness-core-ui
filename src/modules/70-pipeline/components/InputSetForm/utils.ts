/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { SaveToGitFormInterface } from '@modules/10-common/components/SaveToGitForm/SaveToGitForm'
import type { EntityGitDetails } from 'services/pipeline-ng'

export interface GetUpdatedGitDetailsReturnType extends EntityGitDetails {
  lastObjectId?: string
  lastCommitId?: string
  baseBranch?: string
}

export const getUpdatedGitDetails = (
  isEdit: boolean,
  gitDetails: SaveToGitFormInterface | undefined,
  lastObjectId: string,
  initialGitDetails: EntityGitDetails,
  conflictCommitId?: string
): GetUpdatedGitDetailsReturnType => {
  let updatedGitDetails: GetUpdatedGitDetailsReturnType = {}
  if (gitDetails) {
    updatedGitDetails = { ...gitDetails }
    if (isEdit) {
      updatedGitDetails['lastObjectId'] = lastObjectId
      updatedGitDetails['lastCommitId'] = conflictCommitId || initialGitDetails.commitId
    }
    if (gitDetails.isNewBranch) {
      updatedGitDetails['baseBranch'] = initialGitDetails.branch
    }
  }
  return updatedGitDetails
}
