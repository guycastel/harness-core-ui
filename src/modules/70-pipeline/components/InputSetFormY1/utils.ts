/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { cloneDeep, defaultTo, forOwn, pick } from 'lodash-es'
import { CompletionItemKind } from 'vscode-languageserver-types'
import { GitDetails } from '@harnessio/react-pipeline-service-client'
import { CompletionItemInterface } from '@modules/10-common/interfaces/YAMLBuilderProps'
import { EntityGitDetails } from 'services/pipeline-ng'
import { StoreMetadata } from '@modules/10-common/constants/GitSyncTypes'
import {
  FromikInputSetY1,
  InputSetKVPairs,
  InputSetMetadataY1,
  InputSetY1,
  InputSetsY1,
  SelectedItemsType
} from './types'
import { SelectedInputSetListValue } from '../InputSetSelector/SelectedInputSetList'

export const inputSetIdsToSelectedItems = (
  inputSetIds: InputSetsY1,
  allInputSetItems: SelectedItemsType[]
): SelectedItemsType[] => {
  return allInputSetItems.filter(item => inputSetIds.includes(item.value as string))
}

export const getInputSetFromYaml = (
  values: InputSetY1,
  options: { escapeEmpty: boolean } = { escapeEmpty: false }
): InputSetY1 => {
  const ret = cloneDeep(values)

  if (ret.spec) {
    const spec = ret.spec
    if (options.escapeEmpty) {
      forOwn(spec, (value: unknown, key: unknown) => {
        if (key === 'input_sets') return

        if (typeof key === 'string' && typeof value === 'string') {
          spec[key as string] = value === '' ? '""' : value
        }
      })
    }
  }

  return ret
}

export const replaceEmptyWithNull = (values: FromikInputSetY1): { changed: boolean; values: FromikInputSetY1 } => {
  let changed = false
  const retValues = cloneDeep(values)

  forOwn(values.spec, (value: unknown, key: unknown) => {
    if (key === 'input_sets') return

    if (typeof key === 'string') {
      if (retValues.spec) {
        if (value === '' || value === null) {
          changed = true
          retValues.spec[key as string] = null
        } else {
          retValues.spec[key as string] = value
        }
      }
    }
  })

  return { changed, values: retValues }
}

export const addRemoveKeysFromInputSet = (
  inputSet: FromikInputSetY1,
  visibleKeys: string[],
  mergedInputSetValues: InputSetKVPairs
): FromikInputSetY1 => {
  const newInputSet = cloneDeep(inputSet)
  const spec = newInputSet.spec ?? {}

  // add
  visibleKeys.forEach(key => {
    // apply merged or null
    if (typeof spec?.[key] === 'undefined') spec[key] = mergedInputSetValues[key] ?? null
  })

  // remove
  Object.keys(spec).forEach(key => {
    if (!visibleKeys.includes(key) && key !== 'input_sets') {
      delete spec[key]
    }
  })

  newInputSet.spec = spec

  return newInputSet
}

export const getDefaultInputSet = (): InputSetY1 => {
  return {
    version: 1,
    kind: 'input-set'
  }
}
export const formikToYaml = (values: FromikInputSetY1, options?: { escapeEmpty?: boolean }): InputSetY1 => {
  const ret = {
    version: 1,
    kind: 'input-set',
    spec: values.spec
  }

  if (ret.spec) {
    const spec = ret.spec
    if (spec && options?.escapeEmpty) {
      forOwn(ret.spec, (value: unknown, key: unknown) => {
        if (typeof key === 'string' && typeof value === 'string') {
          spec[key as string] = value === '""' ? '' : value
        }
      })
    }
  }

  return ret
}

export const formikToMetadata = (values: FromikInputSetY1): InputSetMetadataY1 => {
  return pick(values, 'name', 'identifier', 'description', 'tags')
}

export const formikToGitMetadata = (values: FromikInputSetY1): StoreMetadata => {
  return {
    branch: values.branch,
    connectorRef: values.connectorRef as string,
    repoName: values.repo,
    filePath: values.filePath,
    storeType: values.storeType
  }
}

export const constructInputSetYamlObject = (item: SelectedInputSetListValue): CompletionItemInterface => ({
  label: defaultTo(item.label, ''),
  insertText: defaultTo(item.value as string, ''),
  kind: CompletionItemKind.Field
})

export const gitDetailsV1toV0 = (gitDetails?: GitDetails): EntityGitDetails | undefined => {
  if (gitDetails) {
    // TODO check missing properties
    return {
      branch: gitDetails.branch_name,
      commitId: gitDetails.commit_id,
      filePath: gitDetails.file_path,
      fileUrl: gitDetails.file_url,
      // isHarnessCodeRepo?: ?
      objectId: gitDetails.object_id,
      // parentEntityConnectorRef: ?,
      // parentEntityRepoName: ?
      // repoIdentifier: ?
      repoName: gitDetails.repo_name,
      repoUrl: gitDetails.repo_url
      // rootFolder: ?
    }
  }
}

export const gitDetailsV0toV1 = (gitDetails: EntityGitDetails): GitDetails | undefined => {
  if (gitDetails) {
    // TODO check missing properties
    return {
      branch_name: gitDetails.branch,
      commit_id: gitDetails.commitId,
      file_path: gitDetails.filePath,
      file_url: gitDetails.fileUrl,
      // isHarnessCodeRepo?: ?
      object_id: gitDetails.objectId,
      // parentEntityConnectorRef: ?,
      // parentEntityRepoName: ?
      // repoIdentifier: ?
      repo_name: gitDetails.repoName,
      repo_url: gitDetails.repoUrl
      // rootFolder: ?
    }
  }
}
