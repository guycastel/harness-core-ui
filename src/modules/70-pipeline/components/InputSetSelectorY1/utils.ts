/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { SelectedInputSetListValue } from '../InputSetSelector/SelectedInputSetList'

export const INPUT_SET_SELECTOR_PAGE_SIZE = 100

export const isInputSetItemSelected = (selected: SelectedInputSetListValue[], identifier: string): boolean => {
  return !!selected.find(item => item.value === identifier)
}
