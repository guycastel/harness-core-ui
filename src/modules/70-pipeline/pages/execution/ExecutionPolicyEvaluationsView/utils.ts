import { defaultTo, forEach, includes, map } from 'lodash-es'
import { GovernanceMetadata } from 'services/cd-ng'
import { Evaluation } from 'services/pm'
import { EvaluationStatus, IacmFormattedData } from './types'

const defaultEmpty = (value: string | number | undefined): string => defaultTo(value, '') as string

const formatIacmGovernanceData = (
  data?: Evaluation[]
): {
  iacmData: IacmFormattedData[]
  iacmStatus: EvaluationStatus
} => {
  const iacmData: IacmFormattedData[] = []
  let status: EvaluationStatus = EvaluationStatus.PASS
  const statuses: EvaluationStatus[] = []
  forEach(data, ({ id, details }) =>
    forEach(details, policySet => {
      statuses.push(defaultTo(policySet.status as EvaluationStatus, EvaluationStatus.PASS))
      iacmData.push({
        policySetId: defaultEmpty(policySet.identifier),
        policySetName: defaultEmpty(policySet.name),
        status: defaultEmpty(policySet.status),
        identifier: defaultEmpty(policySet.identifier),
        created: defaultEmpty(policySet.created),
        accountId: defaultEmpty(policySet.account_id),
        orgId: defaultEmpty(policySet.org_id),
        projectId: defaultEmpty(policySet.project_id),
        policyMetadata: map(policySet.details, policy => ({
          id: id,
          policyId: defaultEmpty(policy?.policy?.identifier),
          policyName: defaultEmpty(policy?.policy?.name),
          severity: defaultEmpty(policySet.status),
          denyMessages: policy?.deny_messages,
          status: defaultEmpty(policySet.status),
          identifier: defaultEmpty(policy?.policy?.identifier),
          accountId: defaultEmpty(policySet.account_id),
          orgId: defaultEmpty(policySet.org_id),
          projectId: defaultEmpty(policySet.project_id),
          created: defaultEmpty(policySet.created),
          updated: defaultEmpty(policySet.updated)
        }))
      })
    })
  )

  if (includes(statuses, EvaluationStatus.ERROR)) {
    status = EvaluationStatus.ERROR
  } else if (includes(statuses, EvaluationStatus.WARNING)) {
    status = EvaluationStatus.WARNING
  }
  return { iacmData, iacmStatus: status }
}

export const formatMetaData = (
  isLoading: boolean,
  IACM_ENABLED: boolean,
  governanceMetadata?: GovernanceMetadata,
  data?: Evaluation[]
): GovernanceMetadata => {
  const details: IacmFormattedData[] = []
  let govData = {
    status: EvaluationStatus.PASS,
    id: null,
    details
  }

  if (governanceMetadata) {
    const { details: opaDetails, status: opaStatus, id } = governanceMetadata
    govData = {
      details: opaDetails,
      status: opaStatus,
      id
    }
  }

  const { iacmData, iacmStatus } = formatIacmGovernanceData(data)
  if (IACM_ENABLED && iacmData && !isLoading) {
    govData.details = [...defaultTo(govData.details, []), ...iacmData]
  }

  if (IACM_ENABLED && iacmStatus !== govData.status) {
    if (govData.status === EvaluationStatus.ERROR || iacmStatus === EvaluationStatus.ERROR) {
      govData.status = EvaluationStatus.ERROR
    } else if (govData.status === EvaluationStatus.WARNING || iacmStatus === EvaluationStatus.WARNING) {
      govData.status = EvaluationStatus.WARNING
    } else {
      govData.status = EvaluationStatus.PASS
    }
  }

  return govData
}
