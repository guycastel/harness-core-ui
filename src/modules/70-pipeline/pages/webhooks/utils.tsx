/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { v4 as nameSpace, v5 as uuid } from 'uuid'
import { IconProps, SelectOption } from '@harness/uicore'
import { Scope } from '@common/interfaces/SecretsInterface'
import { RBACError } from '@rbac/utils/useRBACError/useRBACError'
import { StringsMap } from 'stringTypes'
import { StringKeys } from 'framework/strings'

export interface AddWebhookModalData {
  name: string
  identifier: string
  connectorRef: string
  repo: string
  folderPaths: { id: string; value: string }[]
}

export interface NewWebhookModalProps {
  isEdit: boolean
  initialData: AddWebhookModalData
  entityScope?: Scope
  closeModal: () => void
}
export function processFolderPaths(folderPaths: string[]): { id: string; value: string }[] {
  return folderPaths.map(path => ({ id: uuid('', nameSpace()), value: path }))
}

export const initialWebhookModalData = {
  name: '',
  identifier: '',
  connectorRef: '',
  repo: '',
  folderPaths: processFolderPaths([''])
}

export enum STATUS {
  'loading',
  'error',
  'ok'
}

export interface Error {
  message: RBACError
}

export enum WebhookTabIds {
  ListTab = 'ListTab',
  EventsTab = 'EventsTab'
}

export type WebhookEventStatus = 'FAILED' | 'PROCESSING' | 'QUEUED' | 'SKIPPED' | 'SUCCESSFUL'

export const stringsMap: Readonly<Record<WebhookEventStatus, StringKeys>> = {
  FAILED: 'pipeline.executionFilters.labels.Failed',
  PROCESSING: 'pipeline.eventStatus.processing',
  QUEUED: 'pipeline.eventStatus.queued',
  SKIPPED: 'pipeline.testsReports.skipped',
  SUCCESSFUL: 'pipeline.executionFilters.labels.Success'
}

export function getStatusList(getString: (key: keyof StringsMap) => string): SelectOption[] {
  return Object.keys(stringsMap).map(status => {
    return {
      label: getString(stringsMap[status as WebhookEventStatus]),
      value: status
    }
  })
}

export const iconMap: Record<WebhookEventStatus, IconProps> = {
  SUCCESSFUL: { name: 'tick-circle', size: 10 },
  FAILED: { name: 'warning-sign', size: 9 },
  PROCESSING: { name: 'loading', size: 10 },
  QUEUED: { name: 'queued', size: 10 },
  SKIPPED: { name: 'skipped', size: 8 }
}
