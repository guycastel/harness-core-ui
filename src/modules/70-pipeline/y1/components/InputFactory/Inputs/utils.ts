/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { capitalize } from 'lodash-es'

interface ConditionObject {
  status?: string
  condition: string
}

/**
 * Parses condition string from V1 syntax to extract the status and condition.
 * @param {string} conditionString - The input when string.
 * @returns {ConditionObject} The parsed condition object.
 * @example
 * const conditionString = '<+Always> && <+service.name> == "service2"';
 * Output: { status: 'Always' , condition: '<+service.name> == "service2" }
 */
export const parseConditionString = (conditionString: string): ConditionObject => {
  const conditionParts = conditionString.split('&&').map(part => part.trim())

  return {
    status: conditionParts.length > 1 ? conditionParts[0]?.replace('<+', '').replace('>', '').trim() : undefined,
    condition: conditionParts[conditionParts.length > 1 ? 1 : 0]
  }
}

/**
 * Constructs a condition string as per V1 syntax from a ConditionObject.
 * @param {ConditionObject} conditionObject - The condition object.
 * @returns {string} The constructed condition string.
 * @example
 * const conditionObject = {
 *   status: 'Always',
 *   condition: '<+service.name> == "service2"'
 * };
 * Output: '<+Always> && <+service.name> == "service2"'
 */
export const constructConditionString = (conditionObject: ConditionObject): string => {
  if (conditionObject.status) {
    return `<+${conditionObject.status}> && ${conditionObject.condition}`
  } else {
    return conditionObject.condition
  }
}

export const generateReadableLabel = (name = ''): string => {
  return capitalize(name.split('_').join(' '))
}
