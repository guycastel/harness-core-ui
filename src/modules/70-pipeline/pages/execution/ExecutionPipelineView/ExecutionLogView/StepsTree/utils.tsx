/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { defaultTo } from 'lodash-es'
import {
  ExecutionPipelineGroupInfo,
  ExecutionPipelineItem,
  ExecutionPipelineNode
} from '@pipeline/components/ExecutionStageDiagram/ExecutionPipelineModel'
import { ExecutionNode, InterruptEffectDTO } from 'services/pipeline-ng'
import { getExecutionNodeName } from '@pipeline/utils/execUtils'

export function getRetryInterrupts(step: ExecutionPipelineNode<ExecutionNode>): InterruptEffectDTO[] {
  return defaultTo(step?.item?.data?.interruptHistories, defaultTo(step?.group?.data?.interruptHistories, [])).filter(
    row => row.interruptType === 'RETRY'
  )
}

export function getStepDisplayName(
  step: ExecutionPipelineItem<ExecutionNode> | ExecutionPipelineGroupInfo<ExecutionNode>
): string {
  return getExecutionNodeName(step.data) || step.name
}
