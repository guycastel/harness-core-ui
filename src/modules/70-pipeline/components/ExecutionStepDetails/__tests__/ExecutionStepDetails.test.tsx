/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, fireEvent, findAllByText as findAllByTextGlobal, queryByAttribute } from '@testing-library/react'

import { TestWrapper, CurrentLocation } from '@common/utils/testUtils'
import routes from '@common/RouteDefinitions'
import ExecutionContext, { ExecutionContextParams } from '@pipeline/context/ExecutionContext'
import { accountPathProps, executionPathProps, pipelineModuleParams } from '@common/utils/routeUtils'
import { ExecutionNode, useGetExecutionNode } from 'services/pipeline-ng'
import ExecutionStepDetails from '../ExecutionStepDetails'
import data from './data.json'

jest.mock('@common/components/YAMLBuilder/YamlBuilder')
jest.mock('services/pipeline-ng', () => ({
  useGetApprovalInstance: jest.fn(() => ({ data: {}, loading: false })),
  useGetHarnessApprovalInstanceAuthorization: jest.fn(() => ({ data: {}, loading: false })),
  useAddHarnessApprovalActivity: jest.fn(() => ({ mutate: jest.fn() })),
  useGetExecutionNode: jest.fn(() => ({ data: {}, loading: false }))
}))
jest.mock('@common/components/Duration/Duration', () => ({
  Duration() {
    return <div>MOCK DURATION</div>
  }
}))

const aidaMock = {
  loading: false,
  data: {
    data: {
      valueType: 'Boolean',
      value: 'true'
    }
  }
}

jest.mock('services/cd-ng', () => ({
  useGetSettingValue: jest.fn().mockImplementation(() => aidaMock)
}))

const TEST_PATH = routes.toExecutionPipelineView({
  ...accountPathProps,
  ...executionPathProps,
  ...pipelineModuleParams
})

const pathParams = {
  accountId: 'TEST_ACCOUNT_ID',
  orgIdentifier: 'TEST_ORG',
  projectIdentifier: 'TEST_PROJECT',
  pipelineIdentifier: 'TEST_PIPELINE',
  executionIdentifier: 'TEST_EXECUTION',
  module: 'cd',
  source: 'executions',
  stageId: 'selectedStageId'
}

const executionContext: ExecutionContextParams = {
  ...(data as any),
  pipelineStagesMap: new Map(),
  setLogsToken: jest.fn(),
  refetch: jest.fn(),
  addNewNodeToMap: jest.fn()
}

function TestComponent(props: {
  children?: React.ReactNode
  queryParams?: ExecutionContextParams['queryParams']
  nodesMap?: ExecutionContextParams['allNodeMap']
  selectedStep?: string
}): React.ReactElement {
  const { children, queryParams = {}, nodesMap, selectedStep = '', ...rest } = props
  const [allNodeMap, setAllNodeMap] = React.useState(nodesMap || executionContext.allNodeMap)

  function addNewNodeToMap(uuid: string, node: ExecutionNode): void {
    setAllNodeMap(old => ({
      ...old,
      [uuid]: node
    }))
  }

  return (
    <TestWrapper path={TEST_PATH} pathParams={pathParams}>
      <ExecutionContext.Provider
        value={{ ...executionContext, queryParams, allNodeMap, addNewNodeToMap, selectedStepId: selectedStep }}
      >
        <ExecutionStepDetails {...rest} />
        {props.children}
      </ExecutionContext.Provider>
    </TestWrapper>
  )
}

describe('<ExecutionStepDetails /> tests', () => {
  test('renders normal step', () => {
    const { getByText } = render(<TestComponent selectedStep="normalStep" />)
    expect(getByText('MOCK DURATION')).toBeInTheDocument()
    expect(getByText('execution.stepLogs')).toBeInTheDocument()
  })

  describe('retried steps', () => {
    test('shows step selection', async () => {
      const { getByTestId } = render(
        <TestComponent selectedStep="retriedStep">
          <CurrentLocation />
        </TestComponent>
      )

      const loc1 = getByTestId('location')

      expect(loc1.innerHTML).toEqual(
        '/account/TEST_ACCOUNT_ID/cd/orgs/TEST_ORG/projects/TEST_PROJECT/pipelines/TEST_PIPELINE/executions/TEST_EXECUTION/pipeline'
      )

      const retryLogs1 = getByTestId('retry-logs')

      fireEvent.click(retryLogs1)

      const retries = await findAllByTextGlobal(document.body, 'pipeline.execution.retryStepCount', {
        selector: '.bp3-menu-item > div'
      })

      fireEvent.click(retries[0])

      const loc2 = getByTestId('location')
      expect(loc2.innerHTML).toEqual(
        '/account/TEST_ACCOUNT_ID/cd/orgs/TEST_ORG/projects/TEST_PROJECT/pipelines/TEST_PIPELINE/executions/TEST_EXECUTION/pipeline?retryStep=retryId_1'
      )

      const retryLogs2 = getByTestId('retry-logs')

      fireEvent.click(retryLogs2)

      const current = await findAllByTextGlobal(document.body, 'pipeline.execution.retryStepCount', {
        selector: '.bp3-menu-item > div'
      })

      fireEvent.click(current[current.length - 1])

      const loc3 = getByTestId('location')

      expect(loc3.innerHTML).toEqual(
        '/account/TEST_ACCOUNT_ID/cd/orgs/TEST_ORG/projects/TEST_PROJECT/pipelines/TEST_PIPELINE/executions/TEST_EXECUTION/pipeline'
      )
    })

    // define data outside to avoid infinite loop in react useEffect. It will keep the same reference.
    const responseData = {
      data: {
        uuid: 'retryId_1',
        name: 'Retried Step 1'
      }
    }

    test('fetches data for retry step, if not present', async () => {
      ;(useGetExecutionNode as jest.Mock).mockImplementation(() => ({
        data: responseData,
        loading: false
      }))
      const { container } = render(
        <TestComponent selectedStep="retriedStep" queryParams={{ retryStep: 'retryId_1' }} />
      )
      const step = queryByAttribute('data-name', container, 'Retried Step 1')
      expect(step).toBeInTheDocument()
    })

    test('does not fetches data for retry step, if already present', async () => {
      ;(useGetExecutionNode as jest.Mock).mockImplementation(() => ({
        data: null,
        loading: false
      }))
      const { container } = render(
        <TestComponent
          selectedStep="retriedStep"
          queryParams={{ retryStep: 'retryId_1' }}
          nodesMap={{
            ...executionContext.allNodeMap,
            retryId_1: {
              uuid: 'retryId_1',
              name: 'Already_Present_Data'
            }
          }}
        />
      )
      const step = queryByAttribute('data-name', container, 'Already_Present_Data')
      expect(step).toBeInTheDocument()
    })

    test('shows loader while loading', () => {
      ;(useGetExecutionNode as jest.Mock).mockImplementation(() => ({
        data: null,
        loading: true
      }))
      const { container } = render(<TestComponent selectedStep="retriedStep" />)
      expect(container).toMatchSnapshot()
    })
  })
})
