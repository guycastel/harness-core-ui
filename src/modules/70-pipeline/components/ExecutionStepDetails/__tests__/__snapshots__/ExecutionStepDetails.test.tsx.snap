// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExecutionStepDetails /> tests retried steps shows loader while loading 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <div
        class="title"
      >
        <span>
          common.stepName
        </span>
      </div>
      <div
        class="actions"
      >
        <div
          class="bp3-popover-wrapper"
        >
          <div
            class="bp3-popover-target"
          >
            <button
              class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main retry Button--with-current-color Button--iconOnly Button--without-shadow Button--withRightIcon"
              data-status=""
              data-testid="retry-logs"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                pipeline.execution.retryStepCount
              </span>
              <span
                class="bp3-icon bp3-icon-chevron-down StyledProps--main retryStatusIcon"
                icon="chevron-down"
              >
                <svg
                  data-icon="chevron-down"
                  height="12"
                  viewBox="0 0 16 16"
                  width="12"
                >
                  <desc>
                    chevron-down
                  </desc>
                  <path
                    d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
        <div
          class="toggle"
        >
          <label
            class="label"
            title="pipeline.execution.layouts.right"
          >
            <input
              checked=""
              name="layout"
              type="radio"
              value="RIGHT"
            />
            <span
              class="bp3-icon StyledProps--main"
              data-icon="layout-right"
            >
              <svg
                height="16"
                viewBox="0 0 4.233 4.233"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  transform="translate(0 .423) scale(.21167)"
                >
                  <rect
                    fill="#f2f2f2"
                    height="15"
                    rx="0.5"
                    stroke="#9ea9b4"
                    width="19"
                    x="0.5"
                    y="0.5"
                  />
                  <path
                    d="M12 1h7v14h-7z"
                    fill="#9ea9b4"
                  />
                </g>
              </svg>
            </span>
          </label>
          <label
            class="label"
            title="pipeline.execution.layouts.bottom"
          >
            <input
              name="layout"
              type="radio"
              value="BOTTOM"
            />
            <span
              class="bp3-icon StyledProps--main"
              data-icon="layout-bottom"
            >
              <svg
                height="16"
                viewBox="0 0 4.233 4.233"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill="none"
                  transform="matrix(.21166 0 0 .21166 0 .423)"
                >
                  <rect
                    fill="#f2f2f2"
                    height="15"
                    rx="0.5"
                    stroke="#9ea9b4"
                    width="19"
                    x="0.5"
                    y="0.5"
                  />
                  <path
                    d="M1 10h18v5H1z"
                    fill="#9ea9b4"
                  />
                </g>
              </svg>
            </span>
          </label>
          <label
            class="label"
            title="pipeline.execution.layouts.float"
          >
            <input
              name="layout"
              type="radio"
              value="FLOATING"
            />
            <span
              class="bp3-icon StyledProps--main"
              data-icon="layout-float"
            >
              <svg
                height="16"
                viewBox="0 0 4.233 4.233"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill="none"
                  transform="translate(0 .403) scale(.20982)"
                >
                  <rect
                    fill="#000"
                    fill-opacity="0.05"
                    height="16.333"
                    rx="1"
                    width="20.176"
                  />
                  <rect
                    height="15.333"
                    rx="0.5"
                    stroke="#66798b"
                    stroke-opacity="0.6"
                    width="19.176"
                    x="0.5"
                    y="0.5"
                  />
                  <g
                    fill="#9ea9b4"
                  >
                    <path
                      d="M9.303 3a1 1 0 011-1h7a1 1 0 011 1v9a1 1 0 01-1 1h-7a1 1 0 01-1-1zM14.303 14.5a.5.5 0 01.5-.5h3a.5.5 0 110 1h-3a.5.5 0 01-.5-.5z"
                    />
                  </g>
                </g>
              </svg>
            </span>
          </label>
          <label
            class="label"
            title="pipeline.execution.layouts.minimize"
          >
            <input
              name="layout"
              type="radio"
              value="MINIMIZE"
            />
            <span
              class="bp3-icon bp3-icon-minus StyledProps--main"
              icon="minus"
            >
              <svg
                data-icon="minus"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  minus
                </desc>
                <path
                  d="M13 7H3c-.55 0-1 .45-1 1s.45 1 1 1h10c.55 0 1-.45 1-1s-.45-1-1-1z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </label>
        </div>
      </div>
    </div>
    <div
      class="bp3-spinner"
    >
      <div
        class="bp3-spinner-animation"
      >
        <svg
          height="50"
          stroke-width="8.00"
          viewBox="1.00 1.00 98.00 98.00"
          width="50"
        >
          <path
            class="bp3-spinner-track"
            d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
          />
          <path
            class="bp3-spinner-head"
            d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
            pathLength="280"
            stroke-dasharray="280 280"
            stroke-dashoffset="210"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
